File: 123-best-time-to-buy-and-sell-stock-iii/123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        length = len(prices)
        #[0,left] maximum profit
        left_profit = [0]*length
        #[left, end] maximum profit
        right_profit = [0]*length
        profit = 0
        if length == 1:
            return 0
        
        for ind, price in enumerate(prices):
            if ind == 0:
                min_price = price
            else:
                min_price = min(min_price,price)
            left_profit[ind] = max(profit, price - min_price)
        max_price = 0
        profit = 0
        for ind in range(length-1,-1,-1):
            price = prices[ind]
            max_price = max(max_price,price)
            profit = max(profit, max_price - price)
            right_profit[ind] = profit
        final = 0
        for ind in range(length):
            if ind + 1< length:
                final = max(final,left_profit[ind] + right_profit[ind+1])
            else:
                final = max(final,left_profit[ind])
        return final
            

File: 123-best-time-to-buy-and-sell-stock-iii/123-best-time-to-buy-and-sell-stock-iii.java
class Solution {
    public int maxProfit(int[] prices) {
    //@input: prices + at most two transactions//是否需要特殊处理，有无越界可能
    //@output: maximum profit
    //@edge case : no
    //@steps:
    //in unified way
        
    int t1Cost = Integer.MAX_VALUE, t2Cost = Integer.MAX_VALUE;
    int t1Profit = 0, t2Profit = 0;
    for(int price : prices){
        t1Cost = Math.min(t1Cost, price);
        t1Profit = Math.max(t1Profit, price - t1Cost);
        //reinvent
        t2Cost = Math.min(t2Cost, price - t1Profit);
        t2Profit = Math.max(t2Profit, price - t2Cost);
    }
        return t2Profit;
    }
}